readme.txt
Adrenaline Cube 2   0.1 BETA
 
  by »Ąрī« (and its subclans), Hypernova and Q009

  Big thanks to everyone who has created content used in this game:
 
   Suicizer (models)
   Greaserpirate (OpenGL coding)
   AC (scrips)
   Nyne (scripts)
   RaZgRiZ (scripts)
   MeatROme (scripts)
   Demosthenes (scripts)
   Pyccna (models)
   (if I used your content and your name is not in this list, please let me know)

  Thanks to my clanmates for their support and work:
  
   »Ąрī« execut4ble (coding)
   »Ąрī« Fantasy (playtesting/mapping)
   [CF]Vienna (playtesting/coding)
   »Ðts« Ak1d (playtesting)
   'eS|Marta (playtesting/mapping)
   »Ąрī« antioxidiz3r (QA and support)
   
Hello everyone and welcome to AC2. There's alot of new stuff here, so take some time to read through this. First off, the game has some cool graphics features that Q009 and I have put quite a bit of work into, and they are not meant to be disabled. This means that the game will stay better looking than most competitive games out there, in which you can decrease graphics significantly to give you a visual advantage. You will find more information about this in an in-game menu. 
  Second, vanilla Sauerbraten maps are not playable in this mod by default (they lack the correct entities), however, we will do our best to keep new maps/map ports coming.
  Third, be aware that alot of aspects of the game are quite unpolished, mostly the bot ai, menus and in-game content.
  Fourth, this game is quite difficult to play and has ALOT of hidden tricks which you will have to find for yourself :) So if you see one of us do something that seems hackish you're probably just watching an advanced player.
  
  
  
  
A list of the current features (SauerEnhanced):

new loading screens
new hud
gui sounds
glow particles
new rain particles
heartbeat and damage screen when health is <=25
new spark particles
new weapon firing effects (muzzle smoke, smoke trails, impact particles & sounds)
new and realistic bullet trails (bullet flares fly through the air)
nearmiss sound if bullets fly near you
new teleport effects
sound when you make a frag, (normal frag, nice-shot frag, team-frag)
lots of under water effects (firing a weapon produces bubbles, underwater weapon sounds have been messed up by me)
profile system which keeps track of many stats
more polished graphics overall



Features(AC2)
occlusion check for bullet flares
colorized hud depending on health
all weapons redone (some with secondaries)
lots of new pickups
entities look more realistic in their enviroment
shield now protects 100% of health
shield chargers: stand in them to regenerate shield
item pickup messages
new frag messages depending on weapon used
zoom key held down to zoom instead of toggle
new blood effects
new explosions
new physics: much harder to use and much faster once you learn it :)
jumping height reduced to be more realistic; make more use of weapon boosts and physics tricks!
no player pain sounds
different playermodels have different die sounds
different gib effects
separate damage screens depending on health
screen jump for certain weapons
hitscan weapons have a basic headshot capability
new item pickup sounds
new weapon sounds
6 maps supported so far (vanilla maps will NOT work with this mod)
new weapon models
new player skins
players get killed if falling too far
new shader which sharpens view (similar to antialiasing)



Known issues:
-bots are behaving badly, I need to fix this :)
-occasionally there will be problems with weapons not respawning (this is pretty rare but should be fixed soon)
-on some maps there may be a problem with spawn points that kills you immediately upon spawning; I've fixed most of this, let me know if there's any more incedents
-pickups not giving you any ammo: this is also a rare server problem (similar to weapons not respawning) which I'll try to fix soon



Things to do/not do:

DO make you own weapon binds, the ones shipped are just the ones I use
DO NOT act like a spambot or troll on the servers




























SauerEnhanced/changelog.txt
v1.5.5 - SVN:
	-Justice Edition port
	-New loading screen
	-Some HUD fixes and tweaks
	-Added sound muters (see Readme.txt)
	-Can now disable panning for mapsounds (see Readme.txt)
	-Fixed weather being culled
	-Added (lastsaid) and (getlastsaid) commands
	-Added entity stacking (see Readme.txt)
	-Added world axes (enabled in outline mode)
------
v1.4.8 - Patch:
	-Show info about checking for updates when loading
	-Fixed glow particles visible thru models
	-More weather improvements
	-Added weapon icons to HUD
	-Some more ambient sounds
	-More fx for more maps (.SE files)
	-Some other HUD rework
	-Minor fixes
------
v1.4.7 - Patch:
	-Rewrote weather system code. Rain and snow won't be laggy anymore.
	-Moved some functions to main menu when connected for easier use.
	-More map effects (SE files).
------
v1.4.6 - Patch:
	-Fixed some vars not saving
	-Colour master modes in server list
------
v1.4.5 - Patch:
	-Allow to disable glow particles
	-Allow to enable old-style rifle trail.
------
v1.4.4 - Patch:
	-Delete unnecessary files after analysys
	-Downloaded files will be saved in Sauerbraten/seDownloads
------
v1.4.3 - BETA:
	Added "Check-For-New-Version".
	Added profile system.
	Added working news system.
	New, realistic bullet trails.
	Screenshots are saved in "screenshots" folder now and movies in "videos".
	Disabled annoying auto-sort for servers. Press "Sort servers" button to resort.
	Added heartbeat sound and bloody-screen when health is less than 26.
	Added bullet nearmiss sounds.
	Improved rain and snow.
	Removed screen fx.
	Remade teleport effect.
	Nicer spark effect.
	Light rays aka "Godrays".
	Full UI rework: HUD, GUI, loading screen and UI sounds.
	Rewrote SE file functions.
	Added frag notify sounds (Normal kill, nice-shot kill, team-kill).
	Added glow particles.
------
v1.2.6 - Update:
	Bases will show their names (capture).
	Enhanced gore.
	Fixed loading other entities than particles and sounds from .se file.
	Menus on top (console text no more will cover it).
	New rocket launcher and pistol soudns.
	Tweaked sparks and made them more glowy.
	Bullets are making bubbles when underwater.
	Different sounds when underwater.
	Added underwater ambient sound.
	Don't play map ambient sounds when underwater.
------
v1.1.6 - Update:
	More smoke effects tweaks.
	More explosion sounds.
	Different frag messages for each weapon.
	Rockets and grenades are making bubbles under water.
	Better glows of some particles when without glare.
	Meat cutting sound when using chainsaw.
	Subtle sound when getting chat message.
	Added rocket flying sound.
	Added gib sounds.
	Quick, real-like explosions.
	Tweaked debris.
	Added /seautosay var to enable/disable auto-says.
------
v1.0.4 - Update:
	New file format .se. Storing additional map entities without affecting original map files.
	Added new ambient sounds - stored in .se files
	Added new particles: snow and rain - stored in .se files
	New menu animation.
	New crosshair.
	Smoothed smoke texture.
	Improved bullet flares.
	Play weapon change sound only when local.
	New explosion sounds: for grenades and rockets.
	Added bumpmaps for some md3s and envmaps.
	Fixed bug with home directory.
	Added 2 new sounds for each gun (except rocket launcher).
	Added big light flare effects for flag gamemodes (CTF, Protect).
	Smoke is getting bigger with time.
	New pistol sounds.
	Some sparks for muzzleflashes.
	New CTF sounds.
	Converted sounds to .ogg.
	New font.
	New loading bar.
------
v0.8.4 - Update:
	Fixed a little rocket flames.
------
v0.8.3 - Update:
	Smoke goes up.
	Fixed spark and smoke particles floating in the same way.
	Another methode to fix the /maxfps 60 bug.
	New particle improvement for rockets: flickering light beam.
	Fixed rifle trail yet again.
	Shorten explosion animation.
	Improved rocket flame particles.
	More explosion improvements.
	Shorten muzzle flash fade time.
	Removed spark splash from shotgun muzzle.
	Tweaked shotgun muzzle flash texture.
	Optimized a little code.
	Darken explosion smoke.
	Added variables and menus so users can disable some features of SauerEnhanced.
	Smoke after explosions is now soft-edged (requires /depthfx turned on).
	Added screen effects (teleport, pickups and player respawn).
	Added particle effects when using QuadDamage.
	New, high-transparency GUI skin.
	Added new bullet impact sounds when using QuadDamage.
	New background and logo.
------
v0.6.3 - Update:
	Fixed smoke flying in random directions.
	Fixed the /maxfps 60 bug.
	Added new smoke, blood and spark particle textures.
	Fixed rifle bullet trail.
------
v0.5.9 - Update:
	Improved blood effects.
	Added teammate name showing.
	Decreased size of bullet flare.
	Added script that will show song name and file name which plays.
------
v0.5.5 - Update:
	Fixed DOS console dosesn't showing text.
------
v0.5.4 - Update:
	Separated reload sounds so you don't hear how's somone reloading.
	Added new jumppad sound.
------
v0.5.2 - Update:
	Improved rifle bullet trail particle effects.
	Fixed player1 entity dosesn't playing all pain sounds.
	Improved a little blood effects.
------
v0.4.9 - Update:
	Tweaked rifle muzzle flash.
	Shorten sparks fade time.
	Added smoke for bullet impact.
	Added sound for bullet impact.
	Tweaked rocket particle effects.
	Included source so MAC and Linux users can use the mod.
	Added particle effects for itmes when respawning: ammo, shields and other power-ups.
------
v0.4.2 - Update:
	Improved explosion particle effects.
	Improved rocket particle effects.
------
v0.4.0 - Update:
	Fixed a little smoke effects.
	Separated Sauerbraten and mod files. Mod is now Standalone.
------
v0.3.8 - Update:
	Added new background textures.
	New GUI texture - beta.
	New load bar - beta.
	Deleted font texture from earlier version.

SauerEnhanced/source/shared/ents.h
// this file defines static map entities ("entity") and dynamic entities (players/monsters, "dynent")
// the gamecode extends these types to add game specific functionality

// ET_*: the only static entity types dictated by the engine... rest are gamecode dependent

enum { ET_EMPTY=0, ET_LIGHT, ET_MAPMODEL, ET_PLAYERSTART, ET_ENVMAP, ET_PARTICLES, ET_SOUND, ET_SPOTLIGHT, ET_GAMESPECIFIC };

struct entity                                   // persistent map entity
{
    vec o;                                      // position
    short attr1, attr2, attr3, attr4, attr5;
    uchar type;                                 // type is one of the above
    uchar reserved;
};

struct entitylight
{
    vec color, dir;
    int millis;

    entitylight() : color(1, 1, 1), dir(0, 0, 1), millis(-1) {}
};

struct extentity : entity                       // part of the entity that doesn't get saved to disk
{
    enum
    {
        F_NOVIS     = 1<<0,
        F_NOSHADOW  = 1<<1,
        F_NOCOLLIDE = 1<<2,
        F_ANIM      = 1<<3
    };

    uchar spawned, inoctanode, visible, flags;  // the only dynamic state of a map entity
    entitylight light;
    extentity *attached;

    extentity() : visible(false), flags(0), attached(NULL) {}
};

#define MAXENTS 10000

//extern vector<extentity *> ents;                // map entities

enum { CS_ALIVE = 0, CS_DEAD, CS_SPAWNING, CS_LAGGED, CS_EDITING, CS_SPECTATOR };

enum { PHYS_FLOAT = 0, PHYS_FALL, PHYS_SLIDE, PHYS_SLOPE, PHYS_FLOOR, PHYS_STEP_UP, PHYS_STEP_DOWN, PHYS_BOUNCE };

enum { ENT_PLAYER = 0, ENT_AI, ENT_INANIMATE, ENT_CAMERA, ENT_BOUNCE };

enum { COLLIDE_AABB = 0, COLLIDE_OBB, COLLIDE_ELLIPSE };

struct physent                                  // base entity type, can be affected by physics
{
    vec o, vel, falling;                        // origin, velocity
    vec deltapos, newpos;                       // movement interpolation
    float yaw, pitch, roll;
    float maxspeed;                             // cubes per second, 100 for player
    int timeinair;
    float radius, eyeheight, aboveeye;          // bounding box size
    float xradius, yradius, zmargin;
    float lastvelx, lastvely; //used for compensating for air friction in x/y
    vec floor;                                  // the normal of floor the dynent is on

    int inwater;
    bool jumping;
    char move, strafe;

    uchar physstate;                            // one of PHYS_* above
    uchar state, editstate;                     // one of CS_* above
    uchar type;                                 // one of ENT_* above
    uchar collidetype;                          // one of COLLIDE_* above

    bool blocked;                               // used by physics to signal ai

    physent() : o(0, 0, 0), deltapos(0, 0, 0), newpos(0, 0, 0), yaw(0), pitch(0), roll(0), maxspeed(100),
               radius(4.1f), eyeheight(16), aboveeye(1), xradius(4.1f), yradius(4.1f), zmargin(0),
               state(CS_ALIVE), editstate(CS_ALIVE), type(ENT_PLAYER),
               collidetype(COLLIDE_ELLIPSE),
               blocked(false)
               { reset(); }

    void resetinterp()
    {
        newpos = o;
        deltapos = vec(0, 0, 0);
    }

    void reset()
    {
    	inwater = 0;
        timeinair = 0;
        strafe = move = 0;
        physstate = PHYS_FALL;
        vel = falling = vec(0, 0, 0);
        floor = vec(0, 0, 1);
    }

    vec feetpos(float offset = 0) const { return vec(o).add(vec(0, 0, offset - eyeheight)); }
    vec headpos(float offset = 0) const { return vec(o).add(vec(0, 0, offset)); }

    bool maymove() const { return timeinair || physstate < PHYS_FLOOR || vel.squaredlen() > 1e-4f || deltapos.squaredlen() > 1e-4f; }
};

enum
{
    ANIM_DEAD = 0, ANIM_DYING, ANIM_IDLE,
    ANIM_FORWARD, ANIM_BACKWARD, ANIM_LEFT, ANIM_RIGHT,
    ANIM_HOLD1, ANIM_HOLD2, ANIM_HOLD3, ANIM_HOLD4, ANIM_HOLD5, ANIM_HOLD6, ANIM_HOLD7, ANIM_HOLD8, ANIM_HOLD9, ANIM_HOLD10, ANIM_HOLD11, ANIM_HOLD12, ANIM_HOLD13, ANIM_HOLD14, ANIM_HOLD15, ANIM_HOLD16,
    ANIM_ATTACK1, ANIM_ATTACK2, ANIM_ATTACK3, ANIM_ATTACK4, ANIM_ATTACK5, ANIM_ATTACK6, ANIM_ATTACK7, ANIM_ATTACK8, ANIM_ATTACK9, ANIM_ATTACK10, ANIM_ATTACK11, ANIM_ATTACK12, ANIM_ATTACK13, ANIM_ATTACK14, ANIM_ATTACK15, ANIM_ATTACK16,
    ANIM_PAIN,
    ANIM_JUMP, ANIM_SINK, ANIM_SWIM,
    ANIM_EDIT, ANIM_LAG, ANIM_TAUNT, ANIM_WIN, ANIM_LOSE,
    ANIM_GUN_IDLE, ANIM_GUN_SHOOT,
    ANIM_VWEP_IDLE, ANIM_VWEP_SHOOT, ANIM_SHIELD, ANIM_POWERUP,
    ANIM_MAPMODEL, ANIM_TRIGGER,
    NUMANIMS
};

static const char * const animnames[] =
{
    "dead", "dying", "idle",
    "forward", "backward", "left", "right",
    "hold 1", "hold 2", "hold 3", "hold 4", "hold 5", "hold 6", "hold 7", "hold 8", "hold 9", "hold 10", "hold 11", "hold 12", "hold 13", "hold 14", "hold 15", "hold 16",
    "attack 1", "attack 2", "attack 3", "attack 4", "attack 5", "attack 6", "attack 7", "attack 8", "attack 9", "attack 10", "attack 11", "attack 12", "attack 13", "attack 14", "attack 15", "attack 16",
    "pain",
    "jump", "sink", "swim",
    "edit", "lag", "taunt", "win", "lose",
    "gun idle", "gun shoot",
    "vwep idle", "vwep shoot", "shield", "powerup",
    "mapmodel", "trigger"
};

#define ANIM_ALL         0x7F
#define ANIM_INDEX       0x7F
#define ANIM_LOOP        (1<<7)
#define ANIM_START       (1<<8)
#define ANIM_END         (1<<9)
#define ANIM_REVERSE     (1<<10)
#define ANIM_DIR         0x780
#define ANIM_SECONDARY   11
#define ANIM_NOSKIN      (1<<22)
#define ANIM_SETTIME     (1<<23)
#define ANIM_FULLBRIGHT  (1<<24)
#define ANIM_REUSE       (1<<25)
#define ANIM_NORENDER    (1<<26)
#define ANIM_RAGDOLL     (1<<27)
#define ANIM_SETSPEED    (1<<28)
#define ANIM_GHOST       (1<<29)
#define ANIM_FLAGS       (0x1FF<<22)

struct animinfo // description of a character's animation
{
    int anim, frame, range, basetime;
    float speed;
    uint varseed;

    animinfo() : anim(0), frame(0), range(0), basetime(0), speed(100.0f), varseed(0) { }

    bool operator==(const animinfo &o) const { return frame==o.frame && range==o.range && (anim&(ANIM_SETTIME|ANIM_DIR))==(o.anim&(ANIM_SETTIME|ANIM_DIR)) && (anim&ANIM_SETTIME || basetime==o.basetime) && speed==o.speed; }
    bool operator!=(const animinfo &o) const { return frame!=o.frame || range!=o.range || (anim&(ANIM_SETTIME|ANIM_DIR))!=(o.anim&(ANIM_SETTIME|ANIM_DIR)) || (!(anim&ANIM_SETTIME) && basetime!=o.basetime) || speed!=o.speed; }
};

struct animinterpinfo // used for animation blending of animated characters
{
    animinfo prev, cur;
    int lastswitch;
    void *lastmodel;

    animinterpinfo() : lastswitch(-1), lastmodel(NULL) {}
};

#define MAXANIMPARTS 3

struct occludequery;
struct ragdolldata;

struct dynent : physent                         // animated characters, or characters that can receive input
{
    bool k_left, k_right, k_up, k_down;         // see input code

    entitylight light;
    animinterpinfo animinterp[MAXANIMPARTS];
    ragdolldata *ragdoll;
    occludequery *query;
    int occluded, lastrendered;

    dynent() : ragdoll(NULL), query(NULL), occluded(0), lastrendered(0)
    {
        reset();
    }

    ~dynent()
    {
#ifndef STANDALONE
        extern void cleanragdoll(dynent *d);
        if(ragdoll) cleanragdoll(this);
#endif
    }

    void stopmoving()
    {
        k_left = k_right = k_up = k_down = jumping = false;
        move = strafe = 0;
    }

    void reset()
    {
        physent::reset();
        stopmoving();
    }

    vec abovehead() { return vec(o).add(vec(0, 0, aboveeye+4)); }
};


